src/codegen/revert.rs

```diff
@@ -0,0 +1,364 @@
+use super::expression::expression;
+use super::Options;
+use super::{
+    cfg::{ControlFlowGraph, Instr},
+    vartable::Vartable,
+};
+
+use crate::codegen::Expression;
+use crate::sema::{
+    ast,
+    ast::{FormatArg, Function, Namespace, Type},
+    file::PathDisplay,
+};
+use crate::Target;
+use parse_display::Display;
+use solang_parser::pt::{CodeLocation, Loc, Loc::Codegen};
+
+/// Corresponds to the error types from the Solidity language.
+///
+/// Marked as non-exhaustive because Solidity may add more variants in the future.
+#[non_exhaustive]
+#[derive(Debug, PartialEq, Clone)]
+pub enum SolidityError {
+    /// Reverts with "empty error data"; stems from `revert()` or `require()` without string arguments.
+    Empty,
+    /// The `Error(string)` selector
+    String(Expression),
+    /// The `Panic(uint256)` selector
+    Panic(PanicCode),
+}
+
+impl SolidityError {
+    /// Return the selector expression of the error.
+    pub fn selector_expression(&self) -> Expression {
+        let selector = match self {
+            Self::Empty => unreachable!("empty return data has no selector"),
+            Self::String(_) => self.selector().into(),
+            Self::Panic(_) => self.selector().into(),
+        };
+
+        Expression::NumberLiteral {
+            loc: Codegen,
+            ty: Type::Bytes(4),
+            value: selector,
+        }
+    }
+
+    /// Return the selector of the error.
+    pub fn selector(&self) -> u32 {
+        match self {
+            Self::Empty => unreachable!("empty return data has no selector"),
+            Self::String(_) => 0x08c379a0u32,
+            Self::Panic(_) => 0x4e487b71u32,
+        }
+    }
+
+    /// ABI encode the selector and any error data.
+    pub fn abi_encode(
+        &self,
+        loc: &Loc,
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> Option<Expression> {
+        match self {
+            Self::Empty => None,
+            Self::String(data) => create_encoder(ns, false)
+                .encode_error_data_const(self.clone())
+                .or_else(|| {
+                    let args = vec![self.selector_expression(), data.clone()];
+                    abi_encode(loc, args, ns, vartab, cfg, false).0.into()
+                }),
+            Self::Panic(_) => create_encoder(ns, false).encode_error_data_const(self.clone()),
+        }
+    }
+}
+
+/// Solidity `Panic` Codes. Source:
+/// https://docs.soliditylang.org/en/v0.8.20/control-structures.html#panic-via-assert-and-error-via-require
+#[derive(Display, Debug, PartialEq, Clone, Copy)]
+#[non_exhaustive]
+#[repr(u8)]
+pub enum PanicCode {
+    Generic = 0x00,
+    Assertion = 0x01,
+    MathOverflow = 0x11,
+    DivisionByZero = 0x12,
+    EnumCastOob = 0x21,
+    StorageBytesEncodingIncorrect = 0x22,
+    EmptyArrayPop = 0x31,
+    ArrayIndexOob = 0x32,
+    OutOfMemory = 0x41,
+    InternalFunctionUninitialized = 0x51,
```

Comment:
We don't have a check for this (we need it though!)

Reply:
True, I added it